lib/SILOptimizer/Mandatory/DiagnoseUnreachable.cpp

```diff
@@ -746,6 +747,108 @@ static void performNoReturnFunctionProcessing(SILModule *M,
   }
 }
 
+static bool hasRecursiveCallInPath(SILBasicBlock &Block) {
+  // Process all the Stmt's in this block to find any calls to FD.
+  for (auto &I : Block) {
+    auto *AI = dyn_cast<ApplyInst>(&I);
+    if (AI && AI->getCalleeFunction()
+        && AI->getCalleeFunction() == Block.getParent())
+      return true;
+
+    if (FullApplySite FAI = FullApplySite::isa(&I)) {
+      auto &M = FAI.getModule();
+      if (auto *CMI = dyn_cast<ClassMethodInst>(FAI.getCallee())) {
+        // Don't touch dynamic dispatch.
+        if (CMI->isVolatile())
+          continue;
+
+        auto ClassType = CMI->getOperand()->getType();
+        if (ClassType.is<MetatypeType>())
+          ClassType = ClassType.getMetatypeInstanceType(M);
+
+        auto *F
+            = M.lookUpFunctionInVTable(ClassType.getClassOrBoundGenericClass(),
```

Comment:
@slavapestov Is weeding out volatile methods/witnesses enough here?

Reply:
Update: Doesn’t matter now that dynamic witness methods are always going through ObjcMethodInst